crates/core/transaction/src/plan/build.rs

```diff
@@ -14,72 +17,35 @@ use crate::{
+    /// [`ActionPlan`]s in the TransactionPlan.
+    /// Arguments:
+    ///     self: `TransactionPlan`
+    ///     fvk: `FullViewingKey`
+    ///     witness_data: `WitnessData`
+    /// Returns: `Transaction`
+    pub fn build_unauth_with_actions(
         self,
-        fvk: &FullViewingKey,
+        mut actions: Vec<Action>,
+        fvk: FullViewingKey,
         witness_data: WitnessData,
-    ) -> Result<UnauthTransaction> {
-        let mut actions = Vec::new();
-        let mut synthetic_blinding_factor = Fr::zero();
-
+    ) -> Result<Transaction> {
         // Add the memo.
         let mut memo: Option<MemoCiphertext> = None;
-        let mut memo_key: Option<PayloadKey> = None;
+        let mut _memo_key: Option<PayloadKey> = None;
         if self.memo_plan.is_some() {
             let memo_plan = self
                 .memo_plan
                 .clone()
                 .ok_or_else(|| anyhow!("missing memo_plan in TransactionPlan"))?;
             memo = memo_plan.memo().ok();
-            memo_key = Some(memo_plan.key);
-        }
-
-        // We build the actions sorted by type, with all spends first, then all
-        // outputs, etc.  This order has to align with the ordering in
-        // TransactionPlan::effect_hash, which computes the auth hash of the
-        // transaction we'll build here without actually building it.
-
-        // Build the transaction's spends.
-        for spend_plan in self.spend_plans() {
-            let note_commitment = spend_plan.note.commit();
-            let auth_path = witness_data
-                .state_commitment_proofs
-                .get(&note_commitment)
-                .context(format!("could not get proof for {note_commitment:?}"))?;
-
-            synthetic_blinding_factor += spend_plan.value_blinding;
-            actions.push(Action::Spend(spend_plan.spend(
-                fvk,
-                [0; 64].into(),
-                auth_path.clone(),
-                witness_data.anchor,
-            )));
-        }
-
-        // Build the transaction's outputs.
-        let dummy_payload_key: PayloadKey = [0u8; 32].into();
-        // If the memo_key is None, then there is no memo, and we populate the memo key
-        // field with a dummy key.
-        for output_plan in self.output_plans() {
-            // Outputs subtract from the transaction's value balance.
-            synthetic_blinding_factor += output_plan.value_blinding;
-            actions.push(Action::Output(output_plan.output(
-                fvk.outgoing(),
-                memo_key.as_ref().unwrap_or(&dummy_payload_key),
-            )));
+            _memo_key = Some(memo_plan.key);
         }
 
         // Build the transaction's swaps.
         for swap_plan in self.swap_plans() {
-            synthetic_blinding_factor += swap_plan.fee_blinding;
-            actions.push(Action::Swap(swap_plan.swap(fvk)));
+            actions.push(Action::Swap(swap_plan.swap(&fvk)));
```

Comment:
Wait, why are we building the action here, shouldn't the `actions` come preconstructed?

Reply:
I haven't handled other actions besides `spend` and `output` actions in [build_action](https://github.com/penumbra-zone/penumbra/blob/tx-decompose/crates/wasm/src/wasm_planner.rs#L123) and [build_unauth](https://github.com/penumbra-zone/penumbra/blob/tx-decompose/crates/core/transaction/src/plan/action.rs#L126) yet. The build method will also need account for these [changes](https://github.com/penumbra-zone/penumbra/blob/tx-decompose/crates/core/transaction/src/plan/build.rs#L295). This is planned on being removed.

Reply:
The other action variants have been handled, and `actions` have been made [immutable](https://github.com/penumbra-zone/penumbra/pull/3289/commits).